packages/flutter_tools/lib/src/asset.dart

```diff
@@ -203,6 +230,24 @@ class _Asset {
 
   @override
   String toString() => 'asset: $assetEntry';
+
+  @override
+  bool operator ==(dynamic other) {
+    if (identical(this, other)) return true;
+    if (other.runtimeType != runtimeType) return false;
```

Comment:
I *think* if you did `if (other is! _Asset) return false;`, then the analyzer should be smart enough to know that `other` is an `_Asset` below, so you wouldn't need the extra variable and cast.